src/plugins/intel_gpu/src/plugin/transformations/indirect_kv_cache.cpp

```diff
@@ -0,0 +1,113 @@
+#include "openvino/op/matmul.hpp"
+#include "openvino/op/read_value.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/pass/pattern/op/or.hpp"
+#include "transformations/utils/utils.hpp"
+
+namespace {
+// same impl as ov::replace node, but w/o outputs count check
+void replace_node_unsafe(const std::shared_ptr<ov::Node>& target, const std::shared_ptr<ov::Node>& replacement) {
+    if (ov::op::util::is_output(target)) {
+        OPENVINO_THROW("Result nodes cannot be replaced.");
+    }
+    for (size_t i = 0; i < target->get_output_size(); i++) {
+        target->output(i).replace(replacement->output(0));
+    }
+
+    replacement->add_node_control_dependents(target);
+    replacement->add_node_control_dependencies(target);
+    target->clear_control_dependents();
+}
+
+}  // namespace
+
+namespace ov {
+namespace intel_gpu {
+
+IndirectKVCache::IndirectKVCache() {
+    using namespace ov::pass::pattern;
+
+    auto beam_idx = wrap_type<ov::op::v0::Parameter>();
+    auto gather_input = wrap_type<ov::intel_gpu::op::ReadValue>();
+    auto axis_const = wrap_type<ov::op::v0::Constant>(
+        ov::op::util::constant_predicate<int64_t>([](const std::vector<int64_t>& value) -> bool {
+            return value.size() == 1 && value[0] == 0;
+        }));
+    auto gather_past = wrap_type<ov::op::v8::Gather>({gather_input, beam_idx, axis_const});
+    auto kv_cache = wrap_type<ov::intel_gpu::op::KVCache>({gather_past, any_input()});
+    auto matmul_0 = wrap_type<ov::intel_gpu::op::Gemm>({kv_cache, any_input()});
+    auto matmul_1 = wrap_type<ov::intel_gpu::op::Gemm>({any_input(), kv_cache});
+    auto matmul = std::make_shared<ov::pass::pattern::op::Or>(OutputVector{matmul_0, matmul_1});
+
+    ov::matcher_pass_callback callback = [=](ov::pass::pattern::Matcher& m) {
+        if (transformation_callback(m.get_match_root())) {
+            return false;
+        }
+        const auto& pattern_map = m.get_pattern_value_map();
+
+        auto kv_cache_node = std::dynamic_pointer_cast<ov::intel_gpu::op::KVCache>(pattern_map.at(kv_cache).get_node_shared_ptr());
+
+        auto beam_idx_node = pattern_map.at(beam_idx).get_node_shared_ptr();
+        auto gather_input_node = pattern_map.at(gather_input).get_node_shared_ptr();
+        auto gather_node = pattern_map.at(gather_past).get_node_shared_ptr();
+        ov::replace_node(gather_node, gather_input_node);
+
+        auto indirect_kv_cache = std::make_shared<op::KVCache>(gather_input_node,
+                                                               kv_cache_node->get_input_node_shared_ptr(1),
+                                                               beam_idx_node,
+                                                               kv_cache_node->get_variable(),
+                                                               kv_cache_node->get_concat_axis(),
+                                                               kv_cache_node->get_gather_axis(),
```

Comment:
nit: maybe we can make it clearer and use "actual" gather's zero axis value instead of using default one from `kv_cache_node`